### ISSUE TITLE
Baseline series should use min/max price range in visible range to calculate gradient stops

### ISSUE THREAD

**Issue by markeasting** [tradingview/lightweight-charts#1005]
**Lightweight Charts Version:** 
3.7.0

**Steps/code to reproduce:**
- Add BaselineSeries
- Set a custom price scale
- Assign scaleMargins

```js
var series = chart.addBaselineSeries({
  priceScaleId: 'my-custom-scale',
  scaleMargins: {
    top: 0.8,
    bottom: 0,
  },
  bottomFillColor2: 'rgba(0,0,0,1)',
  bottomFillColor1: 'rgba(0,0,0,0)',
  topFillColor2: 'rgba(0,0,0,0)',
  topFillColor1: 'rgba(0,0,0,1)',
  baseValue: { type: 'price', price: 180 },
});
```

**Actual behavior:**
The baseline gradient does not fully render the assigned color stop from the configuration. 

**Expected behavior:**
The baseline gradient should use the scaleMargins to compute the color stops. 

```js
// src/renderers/baseline-renderer.ts:21
const gradient = ctx.createLinearGradient(0, 0, 0, data.bottom);
```

`data.bottom` should not be the height of the current chart, but the height of the series with the current scaleMargins. 

**Screenshots:**
Expected here would be a full-black (`#000000`) top and bottom gradient. 

Without scale margins:
![image](https://user-images.githubusercontent.com/11995840/153607812-561e0b05-ae45-4e9f-b602-21a1f25be1d7.png)

Scale margins moved to top: 
![image](https://user-images.githubusercontent.com/11995840/153607442-9b3988d7-c73f-4fe0-b306-9f764461e880.png)

Scale margins moved to bottom:
![image](https://user-images.githubusercontent.com/11995840/153607489-a0109015-195e-4e21-ba0f-6cb94d5d286e.png)

**CodeSandbox/JSFiddle/etc link:**
https://jsfiddle.net/0r3cLmjd/51/

**Comment by timocov** [comment]
`scaleMargins` isn't supposed to be a part of the series, it is an option of the price scale (we're going to remove this property from series' options). Thus it looks like intended behaviour - the library uses the full width and height of the canvas to declare a gradient boundaries.

Could you please elaborate what do you want to achieve with that?

**Comment by EastingAndNorthing** [comment]
What I'd like to see is a consistent visual style for every BaselineSeries I add to my chart, wherever they are on the scale. 

I currently have an upper and lower area for my 2 BaselineSeries, the top one looks different from the bottom one, even though the style options are the same. 

In the meantime, I seem to have partially fixed it by adding a quick hack: 

```js

// baseline-pane-view.ts:renderer()

// Added lines
const baselineProps = this._series.options();
const t = (1 - baselineProps.scaleMargins.top) * height;
const b = (1 - baselineProps.scaleMargins.bottom) * height;

this._baselineAreaRenderer.setData({
  items: this._items,
  
  topFillColor1: baselineProps.topFillColor1,
  topFillColor2: baselineProps.topFillColor2,
  bottomFillColor1: baselineProps.bottomFillColor1,
  bottomFillColor2: baselineProps.bottomFillColor2,
  
  lineWidth: baselineProps.lineWidth,
  lineStyle: baselineProps.lineStyle,
  lineType: LineType.Simple,
  
  baseLevelCoordinate,
  top: t as Coordinate,		// Added line
  bottom: b as Coordinate,	// Added line
  
  visibleRange: this._itemsVisibleRange,
  barWidth,
});
```

```js
// baseline-renderer.ts
protected override _fillStyle(ctx: CanvasRenderingContext2D): CanvasRenderingContext2D['fillStyle'] {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const data = this._data!;
  
  // New lines
  const gradient = ctx.createLinearGradient(0, data.top, 0, data.bottom);
  const base = data.baseLevelCoordinate - data.top;
  const h = data.bottom - data.top;
  const baselinePercent = clamp(base / h, 0, 1);
  
  gradient.addColorStop(0, data.topFillColor1);
  gradient.addColorStop(baselinePercent, data.topFillColor2);
  gradient.addColorStop(baselinePercent, data.bottomFillColor1);
  gradient.addColorStop(1, data.bottomFillColor2);
  
  return gradient;
}

```

Test before (notice lack of upper blue in lower chart and lack of purple in upper chart):
![image](https://user-images.githubusercontent.com/11995840/153628172-3e6dbb61-3cd0-40e7-b4f7-67432d0eadec.png)

Test after:
![image](https://user-images.githubusercontent.com/11995840/153628117-03f2e50f-206a-4322-b63b-79e6feafa1ea.png)


As you can see, both series are now completely covered by their gradients as I would expect.

**Comment by timocov** [comment]
Have you tried your solution in the case when a price scale is manually scaled and you move the chart to vertically (especially out of your margins), how does it work?

**Comment by timocov** [comment]
Probably we need to provide a way to control what the start/end points should be. Right now we use the full height of the pane to paint it so I'm trying to understand what exact options we should have. As one of possible options could be max/min coordinates of a series for this visible range, but we'll need to check it properly.

**Comment by EastingAndNorthing** [comment]
I tried with the following code to change the price scale manually (series initialized without scale beforehand):

```js
chart.priceScale('myscale').applyOptions({
    scaleMargins: {
        top: 0,
        bottom: 0.8
    },
});
```

Before:
![image](https://user-images.githubusercontent.com/11995840/153644864-473fd710-4b99-4b8c-b832-90746b1b355a.png)

After applying price scale options:
![image](https://user-images.githubusercontent.com/11995840/153644978-ad4786ad-a47e-4d15-87cd-ba7bd7e5c4fc.png)

You can see that the gradient has changed, especially noticeable in the yellow area. And the purple on top became more purple (since it is now higher, closer to y=0)

**Comment by EastingAndNorthing** [comment]
Regarding your second comment, I think the most logical situation would be that the gradient is simply based on the min/max values of the series (or well, the top/bottom of the pane). 

Having extra options for the start/end point in the API would seem cluttered, in my opinion 👍

**Comment by EastingAndNorthing** [comment]
Area series also seems to be affected. 

One of the gradient color stops is not reached when using scaleMargins on its priceScale:

Without margins:
<img width="189" alt="image" src="https://user-images.githubusercontent.com/11995840/153713787-20dbf911-e4a9-4e66-a743-943cccd1363d.png">

With margins:
<img width="173" alt="image" src="https://user-images.githubusercontent.com/11995840/153713804-d2410822-d206-49b6-8a11-2278c159dfc7.png">

Suddenly the upper side is not black anymore.

I have a fix for both of these issues, may I create a pull request?

**Comment by timocov** [comment]
> You can see that the gradient has changed, especially noticeable in the yellow area. And the purple on top became more purple

Have you tried to move a series after that vertically? How the color is changing on this?

> Having extra options for the start/end point in the API would seem cluttered, in my opinion 

Hm, probably we need to have 2 separate _hard-coded_ options here:

- if a base value is percent (which is not supported right now, but we're keeping it in mind because it is available on tradingview.com in this way) then we should consider the whole height as a bottom value and 0 a top one
- if a base value is price (this is the only available option right now) then we should use min/max values for bottom/top coordinates of a gradient.

What do you think about that @EastingAndNorthing?

> I have a fix for both of these issues, may I create a pull request?

I'll discuss this question with the team and come back to you with the solution we could accept here.

**Comment by timocov** [comment]
Ok, it seems that we're happy with the solution you provided, i.e. calculate top/bottom based on the data in visible range instead of `0..height` as it is right now. But this looks like a breaking changes so it is going to be released in the next major version v4.

@EastingAndNorthing Do you still want to create a pull request for that? See https://github.com/tradingview/lightweight-charts/blob/master/CONTRIBUTING.md#pull-requests for our contribution guide.

**Comment by EastingAndNorthing** [comment]
> Have you tried to move a series after that vertically? How the color is changing on this?

Yes, this will give the same result, a color change while moving the chart up/down. See https://jsfiddle.net/ean9vj3r/1/ 

> What do you think about that @EastingAndNorthing?

I agree, compared to the baseline series on Tradingview, it does seem more logical to use the entire scale as a gradient if we're talking about percentage values. 

> @EastingAndNorthing Do you still want to create a pull request for that?

Yes. I'll also add a few lines to check whether the base value is price, leaving the upcoming percentage scale at `0..height`.

**Comment by SlicedSilver** [comment]
This was implemented within version 5.
https://tradingview.github.io/lightweight-charts/docs/release-notes#500
